softlearning/algorithms/sac.py

```diff
@@ -233,15 +237,47 @@ def _init_actor_update(self):
             name: self._placeholders['observations'][name]
             for name in self._Qs[0].observation_keys
         }
+
         Q_inputs = flatten_input_structure({
             **Q_observations, 'actions': actions})
         Q_log_targets = tuple(Q(Q_inputs) for Q in self._Qs)
         min_Q_log_target = tf.reduce_min(Q_log_targets, axis=0)
 
+        # TODO: create a discriminator here
+        z_value = self._placeholders['z_value']
+        discriminator_inputs = flatten_input_structure({
+            name: self._placeholders['raw_observations'][name]
+            for name in self._policy.observation_keys
+        })[0]
+
+        model = tf.keras.Sequential()
+        model.add(tf.keras.layers.Dense(self.dis_hidden_layers_sizes[0], input_shape=(int(discriminator_inputs.shape[1]),)))
+        model.add(tf.keras.layers.Dense(self.dis_output_size))
+        discriminator_output = model(discriminator_inputs)
+
+        # TODO: create a custom loss using discriminator
+        normal_dist = tfp.distributions.MultivariateNormalDiag(
+            loc=tf.zeros(self.z_dim),
+            scale_diag=tf.ones(self.z_dim))
+        mean_dist = discriminator_output[:,:self.z_dim]
+        log_scale_dist = discriminator_output[:,self.z_dim:]
+        log_scale_dist = tf.clip_by_value(log_scale_dist, clip_value_min=-10, clip_value_max=2)
+        scale_dist = tf.exp(log_scale_dist)
+        output_dist = tfp.distributions.MultivariateNormalDiag(
+            loc=mean_dist,
+            scale_diag=scale_dist)
+        discriminator_log_probs = output_dist.log_prob(z_value)
+        discriminator_log_probs = tf.clip_by_value(discriminator_log_probs, clip_value_min=-20, clip_value_max=20)
+        normal_log_probs = normal_dist.log_prob(z_value)
+        normal_log_probs = tf.clip_by_value(normal_log_probs, clip_value_min=-20, clip_value_max=20)
```

Comment:
for all these log_probs you might need to add a .sum(dim=-1) (the tensorflow version of that) as for now I think you compute the probability per z-dimension but to get the joint log_prob for the whole z-vector (e.g. 6-dimensional or whatever you choose) you need to sum the logs (i.e. multiply the probabilities)

in the end the log_prob should be one number per batch element